nemo/collections/speechlm/models/speech_to_text_llm_model.py

```diff
@@ -0,0 +1,1127 @@
+                self.validation_step_outputs[dataloader_idx].append(outputs)
+            else:
+                self.validation_step_outputs.append(outputs)
+        else:
+            if self._num_test_dl > 1:
+                self.test_step_outputs[dataloader_idx].append(outputs)
+            else:
+                self.test_step_outputs.append(outputs)
+        return forward_output
+
+    def get_inference_strategy(self):
+        return self.config.text_generation_strategy(self.module)
+
+    def predict_step(self, batch: dict, batch_idx: int, dataloader_idx: Optional[int] = None):
+        """
+        Used to get LLM predictions for validation and test steps based on the given inference config.
+        """
+        inference_config = self.get_inference_config()
+        if inference_config is not None:
+            # need to overwrite some configuration, make it immutable
+            inference_config = inference_config.copy()
+        else:
+            self.set_inference_config(inference_config=default_inference_config)
+            logging.warning(f'inference_config is not set. Use default: {default_inference_config}')
+            inference_config = self.get_inference_config()
+
+        if self.cfg.data.get('end_string', None):
+            inference_config['end_strings'] = [self.cfg.data.end_string]
+
+        inference_config['strategy'] = self.get_inference_strategy()
+
+        global_batch_size_per_gpu = batch['tokens'].size(0)
+        num_micro_batches_before_decode = get_num_microbatches()
+
+        compute_logprob = inference_config.get('compute_logprob', False)
+
+        if compute_logprob:
+            inference_config['inputs'] = batch
+            inference_config['tokens_to_generate'] = 1
+            inference_config['all_probs'] = True
+            inference_config['add_BOS'] = False
+            inference_config['greedy'] = True
+            response = generate(self, **inference_config)
+            response = get_computeprob_response(self.tokenizer, response, batch)
+        else:
+            if isinstance(batch, list):
+                inference_config['inputs'] = batch
+            elif 'num_audios' in batch:
+                inference_config['inputs'] = (
+                    batch['contexts'].cuda(),
+                    batch['context_lengths'].cuda(),
+                    batch['audio_signal'].cuda(),
+                    batch['audio_signal_length'].cuda(),
+                    batch['num_audios'].cuda(),
+                    batch['context_start_idx'],
+                )
+            else:
+                inference_config['inputs'] = (
+                    batch['contexts'].cuda(),
+                    batch['context_lengths'].cuda(),
+                    batch['audio_signal'].cuda(),
+                    batch['audio_signal_length'].cuda(),
+                )
+            response = generate(self, **inference_config)
+
+        app_state = AppState()
+        reconfigure_num_microbatches_calculator(
```

Comment:
Why do we need to reconfigure micro batch size on the fly ? Document in one-two sentences

Reply:
I think it's only needed when the last batch has different batch size from the previous batches. This line was copied from nemo1.x implementation of speechlm.